File: cos/olympus/src/main/java/cos/olympus/game/Spells.java
package cos.olympus.game;

import cos.logging.Logger;
import cos.olympus.game.events.Damage;
import cos.olympus.game.events.Fireball;
import cos.olympus.game.events.Shot;
import cos.olympus.util.OpConsumer;
import cos.ops.in.FireballEmmit;
import cos.ops.out.FireballMoved;
import cos.ops.in.MeleeAttack;
import cos.ops.out.MeleeAttacked;
import cos.ops.in.ShotEmmit;
import cos.ops.out.ShotMoved;

import java.util.ArrayList;

import static cos.olympus.util.TimeUtil.toTickSpeed;
import static cos.olympus.util.TimeUtil.toTicks;

public class Spells {
    private final static Logger logger = Logger.get(Spells.class);

    protected static int SPELL_IDS = 0;
    private          int pause     = toTicks(1);

    private final ArrayList<SpellStrategy> spells = new ArrayList<>();
    private final World                    world;

    public Spells(World world) {
        this.world = world;
    }

    void onShot(int tick, ShotEmmit op) {
        var cr = world.getCreature(op.userId());
        if (cr == null) return;
        //todo validate cooldown

        if (tick - cr.lastSpellTick < pause) {
            logger.info("Ignore spell " + op);
            return;
        }

        cr.lastSpellTick = tick;

        var spell = new Shot(++SPELL_IDS, cr.x(), cr.y(), toTickSpeed(1000), cr.sight(), 10, tick, cr);
        var str = new ShotSpellStrategy(spell, world);
        spells.add(str);
    }

    void onSpell(int tick, FireballEmmit op) {
        var cr = world.getCreature(op.userId());
        if (cr == null) return;
        //todo validate cooldown

        if (tick - cr.lastSpellTick < pause) {
            logger.info("Ignore spell " + op);
            return;
        }

        cr.lastSpellTick = tick;

        var spell = new Fireball(++SPELL_IDS, cr.x(), cr.y(), toTickSpeed(400), cr.sight(), 8, tick, cr);
        var str = new FireballSpellStrategy(spell, world);
        spells.add(str);
    }

    void onMeleeAttack(int tick, MeleeAttack op) {
        var cr = world.getCreature(op.userId());
        if (cr == null) return;
        if (tick - cr.lastSpellTick < pause) {
            return;
        }


        var spell = new cos.olympus.game.events.MeleeAttack(++SPELL_IDS, tick, cr.x(), cr.y(), cr.sight(), cr);
        var str = new MeleeAttackStrategy(spell, world);
        spells.add(str);
    }

    public void onTick(int tick, ArrayList<Damage> damages, OpConsumer outOps) {
        spells.forEach(s -> s.onTick(tick, damages));

        spells.forEach((SpellStrategy strategy) -> {
            var spell = strategy.spell();
            world.getAllCreatures().forEach(cr -> {
                if (strategy.inZone(cr)) {
                    if (cr.zoneSpells.put(strategy.id(), strategy) == null) {
                        if (spell instanceof Fireball s) {
                            outOps.add(new FireballMoved(SPELL_IDS++, tick, cr.id(), s.id(), s.x(), s.y(), s.speed(), s.dir(), strategy.isFinished()));
                        } else if (spell instanceof cos.olympus.game.events.MeleeAttack s) {
                            outOps.add(new MeleeAttacked(SPELL_IDS++, tick, cr.id(), s.id(), s.source().id()));
                        } else if (spell instanceof Shot s) {
                            outOps.add(new ShotMoved(SPELL_IDS++, tick, cr.id(), s.id(), s.x(), s.y(), s.speed(), s.dir(), strategy.isFinished()));
                        }
                    }
                }
            });
        });
    }

    public void onAfterTick() {
        spells.removeIf(SpellStrategy::isFinished);
    }
}

File: cos/src/main/java/cos/olympus/game/Spells.kt
package cos.olympus.game

import cos.logging.Logger
import cos.olympus.game.events.Fireball
import cos.olympus.game.events.MeleeAttack
import cos.olympus.game.events.Shot
import cos.olympus.game.events.Spell
import cos.olympus.game.strategy.FireballSpellStrategy
import cos.olympus.game.strategy.MeleeAttackStrategy
import cos.olympus.game.strategy.ShotSpellStrategy
import cos.olympus.game.strategy.SpellStrategy
import cos.olympus.util.OpConsumer
import cos.olympus.util.TimeUtil
import cos.ops.`in`.FireballEmmit
import cos.ops.`in`.ShotEmmit

class Spells(private val world: World) : TickAware {
    private var tick: Int = 0
    private val pause = TimeUtil.toTicks(1)
    private val spells = ArrayList<SpellStrategy>()

    override fun onTick(tick: Int) {
        this.tick = tick
    }

    fun onShot(op: ShotEmmit) {
        val a = world.getActor(op.userId) ?: return

        //todo validate cooldown
        if (tick - a.lastSpellTick < pause) {
            logger.info("Ignore spell $op")
            return
        }

        a.lastSpellTick = tick

        val spell = Shot(++SPELL_IDS, a.x, a.y, TimeUtil.toTickSpeed(1000), a.sight, 10, tick, a)
        val str = ShotSpellStrategy(spell, world)
        spells.add(str)
    }

    fun onSpell(op: FireballEmmit) {
        val a = world.getActor(op.userId) ?: return

        //todo validate cooldown
        if (tick - a.lastSpellTick < pause) {
            logger.info("Ignore spell $op")
            return
        }

        a.lastSpellTick = tick

        val spell = Fireball(++SPELL_IDS, a.x, a.y, TimeUtil.toTickSpeed(400), a.sight, 8, tick, a)
        val str = FireballSpellStrategy(spell, world)
        spells.add(str)
    }

    fun onMeleeAttack(op: cos.ops.`in`.MeleeAttack) {
        val a = world.getActor(op.userId) ?: return
        if (tick - a.lastSpellTick < pause) {
            //too fast
            return
        }


        val spell = MeleeAttack(++SPELL_IDS, tick, a.x, a.y, a.sight, a)
        val str = MeleeAttackStrategy(spell, world)
        spells.add(str)
    }

    fun onMeleeAttack(a: Actor) {
        val spell = MeleeAttack(++SPELL_IDS, tick, a.x, a.y, a.sight, a)
        val str = MeleeAttackStrategy(spell, world)
        spells.add(str)
    }

    fun onTick(damages: Damages, outOps: OpConsumer) {
        spells.forEach { it.onTick(tick, damages) }

        //notify
        spells.forEach { strategy ->
            val spell: Spell = strategy.spell
            world.allActors.forEach { a ->
                if (strategy.inZone(a)) {
                    if (a.zoneSpells.put(strategy.id, strategy) == null) {
                        when (spell) {
                            is Fireball -> {
                                outOps.add(cos.ops.out.Fireball(SPELL_IDS++, tick, a.id, spell.id, spell.x, spell.y, spell.speed, spell.dir, strategy.finished))
                            }

                            is MeleeAttack -> {
                                outOps.add(cos.ops.out.MeleeAttack(SPELL_IDS++, a.id, spell.id, spell.source.id))
                            }

                            is Shot -> {
                                outOps.add(cos.ops.out.Shot(SPELL_IDS++, a.id, spell.id, spell.x, spell.y, spell.speed, spell.dir, strategy.finished))
                            }
                        }
                    }
                }
            }
        }
    }

    fun onAfterTick() {
        spells.removeIf { obj: SpellStrategy -> obj.finished }
    }

    companion object {
        private val logger: Logger = Logger.get(Spells::class.java)

        protected var SPELL_IDS: Int = 0
    }
}
